keps/sig-storage/20200206-namespace-transfer.md

```diff
@@ -0,0 +1,217 @@
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+This KEP proposes an API to securely transfer Kubernetes stroage resources between namespaces.
+It will focus mainly on PersistentVolumeClaims but could apply to VolumeSnapshots as well.
+
+`StorateTransferRequest` resources are created in the target namespace.  Corresponding 
+`StorageTransferApproval` resources are created in the source namespace.  When a controller
+detects matching request/approval resources, a transfer in ititiated.
+
+When transferring a PersistentVolumeClaim, the corresponding PersistentVolume will be updated, 
+the source PersistentVolumeClaim will be deleted, and the target PersistentVolumeClaim will be created 
+(not necesserarily in that order).  The transfer simply deals with Kubernetes API resources.
+No data on the physical PersistentVolume is accessed/copied. 
+
+## Motivation
+
+Give Kubernetes users the flexability to easily and securely share data between namespaces.  
+
+### Goals
+
+- Define an API for transferring persistent storage resources between namespaces
+- The API should be compatible with as many existing provisioner implementations as possible
+
+### Non-Goals
+
+- Transfer of resources other than PersistentVolumeClaims and VolumeSnapshots will not be discussed
+- Implementation details only discussed in the context of influencing the API definition
+
+## Proposal
+
+`StorageTransferRequest` resource:
+
+```yaml
+apiVersion: v1alpha1
+kind: StorageTransferRequest
+metadata:
+  name: transfer-foo
+  namespace: target-namespace
+spec:
+  source:
+    name: foo
+    namespace: source-namespace
+    kind: PersistentVolumeClaim
+  targetName: bar #this is optional
+```
+
+`StorageTransferApproval` resource:
+
+```yaml
+apiVersion: v1alpha1
+kind: StorageTransferApproval
+metadata:
+  name: approve-foo
+  namespace: source-namespace
+spec:
+  source:
+    name: foo
+    kind: PersistentVolumeClaim
+  targetNamespace: target-namespace
+```
+
+Once a `StorageTransferRequest` and corresponding `StorageTransferApproval` are created,
+a controller will begin the transfer process which does the following (not necesserarily in order):
+
+- Make sure the associated PersistentVolume is not deleted/recycled during the transfer by setting reclaim to `Reclaim`.
+- Delete the PVC `source-namespace\foo`
+- Bind The PersistentVolume to `target-namespace\bar`
+- Create the PVC `target-namespace\bar` with the `spec.volumeName` set appropriately
+
+The last step makes it possible for the API to be compatible with the most provisioners
+as they will ignore a PVC that is already bound.  In previous namespace transfer KEPs, 
+the target PVC is created by the user.  Although, this will probably work with the CSI
```

Comment:
Yeah, I think I like your approach better; it sort of fits better, collapsing the pvc-create into the transfer-approval.  I think the only reason I did it that way was that I required the request to have a "transfer-to" name that matched on the other side, but I think that was because of a limitation that I overcame in subsequent changes.

Anyway, long winded way of saying I think this is a better model.